ga4gh/datasource/htslib.py

```diff
@@ -0,0 +1,915 @@
+                attrName, attr, minVal, maxVal))
+        if attr > maxVal:
+            raise exceptions.DatamodelValidationException(message.format(
+                attrName, attr, minVal, maxVal))
+
+    @classmethod
+    def assertInt(cls, attr, attrName):
+        if not isinstance(attr, (int, long)):
+            message = "invalid {} '{}' not an int".format(attrName, attr)
+            raise exceptions.DatamodelValidationException(message)
+
+    @classmethod
+    def sanitizeInt(cls, attr, minVal, maxVal, attrName):
+        cls.assertInt(attr, attrName)
+        if attr < minVal:
+            attr = minVal
+        if attr > maxVal:
+            attr = maxVal
+        return attr
+
+    @classmethod
+    def sanitizeString(cls, attr, attrName):
+        if not isinstance(attr, basestring):
+            message = "invalid {} '{}' not a string".format(
+                attrName, attr)
+            raise exceptions.DatamodelValidationException(message)
+        if isinstance(attr, unicode):
+            attr = attr.encode('utf8')
+        if len(attr) > cls.maxStringLength:
+            attr = attr[:cls.maxStringLength]
+        return attr
+
+    def get_file_handle(self, open_func, url, *args, **kwargs):
+        return file_handle_cache.get_file_handle(
+            url, open_func, url, *args, **kwargs)
+
+
+class HtslibReference(registry.Reference, PysamMixin):
+    __tablename__ = 'HtslibReference'
+
+    id = sqlalchemy.Column(
+        sqlalchemy.Integer, sqlalchemy.ForeignKey('Reference.id'),
+        primary_key=True)
+    fasta_url = sqlalchemy.Column(sqlalchemy.String, nullable=False)
+    __mapper_args__ = {
+        'polymorphic_identity': 'HtslibReference',
+    }
+
+    def run_get_bases(self, start, end):
+        self.check_query_range(start, end)
+        fasta_file = self.get_file_handle(pysam.FastaFile, self.fasta_url)
+        bases = fasta_file.fetch(self.name.encode(), start, end)
+        return bases
+
+
+class HtslibReferenceSet(registry.ReferenceSet):
+    __tablename__ = 'HtslibReferenceSet'
+    id = sqlalchemy.Column(
+        sqlalchemy.Integer, sqlalchemy.ForeignKey('ReferenceSet.id'),
+        primary_key=True)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'HtslibReferenceSet',
+    }
+
+    def __init__(self, name, fasta_url):
+        super(HtslibReferenceSet, self).__init__(name)
+        with pysam.FastaFile(fasta_url) as fasta_file:
+            for reference_name in fasta_file.references:
+                reference = HtslibReference(reference_name)
+                bases = fasta_file.fetch(reference_name)
+                reference.md5checksum = hashlib.md5(bases).hexdigest()
+                reference.length = len(bases)
+                reference.fasta_url = fasta_url
+                self.references.append(reference)
+        # Now that we have all the references, set the md5checksum
+        self._compute_md5checksum()
+
+##################################################
```

Comment:
I was thinking the same thing. There's still some work to be done to factor out some common code, but I definitely agree that this should be split up once it's fairly solid and stable.